### PR TITLE
Allow manifest rebuild to loop up to 100 times

### PR DESCRIPTION
**Pull request by colincross**
Ninja generators that bootstrap themselves with Ninja may need
to rebuild build.ninja multiple times.  Replace the 2 cycle loop
with a 100 cycle loop, and print the pass number each time it
restarts.

As discussed in https://groups.google.com/forum/#!searchin/ninja-build/jamie/ninja-build/YHOpOIS3FSo/ProjUh-DUxkJ

### CODE REVIEW

File: src/ninja.cc [hunk 1]
```diff
@@ -1061,9 +1061,11 @@ int real_main(int argc, char** argv) {
     return (ninja.*options.tool->func)(argc, argv);
   }
 
-  // The build can take up to 2 passes: one to rebuild the manifest, then
-  // another to build the desired target.
-  for (int cycle = 0; cycle < 2; ++cycle) {
+  // The build can take up to 100 passes to rebuild the manifest and then to
```
Comment (nico) [inline src/ninja.cc]: `// Limit number of rebuilds, to prevent infinite loops.`

File: src/ninja.cc [hunk 2]
```diff
@@ -1061,9 +1061,11 @@ int real_main(int argc, char** argv) {
...
+  // build the desired target, far more times than expected in any normal
+  // use case, but will prevent infinite loops.
+  const int numPasses = 100;
```
Comment (nico) [inline src/ninja.cc]: `kCycleLimit`

File: src/ninja.cc [hunk 3]
```diff
@@ -1086,16 +1088,17 @@ int real_main(int argc, char** argv) {
...
-        // Start the build over with the new manifest.
-        continue;
-      } else if (!err.empty()) {
-        Error("rebuilding '%s': %s", options.input_file, err.c_str());
-        return 1;
+    // Attempt to rebuild the manifest before building anything else
+    if (ninja.RebuildManifest(options.input_file, &err)) {
+      // Start the build over with the new manifest.
+      if (cycle < numPasses) {
+        printf("%s updated, restarting ninja for pass %d\n",
```
Comment (nico) [inline src/ninja.cc]: Remove ("Developers should design programs so that they do not print unnecessary output" etc)

File: src/ninja.cc [hunk 4]
```diff
@@ -1086,16 +1088,17 @@ int real_main(int argc, char** argv) {
...
+               options.input_file, cycle + 1);
       }
+      continue;
+    } else if (!err.empty()) {
+      Error("rebuilding '%s': %s", options.input_file, err.c_str());
```
Comment (nico) [inline src/ninja.cc]: if it's important in your experience, log the cycle here (`"rebuildling '%s' (cycle %d): …`)

**Comment by nico** [comment]
Thanks! Couple nits, else looks good.

**Comment by colincross** [comment]
Done

**Pull request merged**